src/mlpack/tests/main_tests/kfn_test.cpp

```diff
@@ -0,0 +1,378 @@
+
+  // Random input, some k > number of reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("query", std::move(queryData));
+  SetInputParam("k", (int) 101);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Check that we can't specify a negative leaf size.
+ */
+BOOST_AUTO_TEST_CASE(KFNLeafSizeTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+  
+  // Random input, negative leaf size.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("leaf_size", (int) -1); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/*
+ * Check that we can't pass both input_model and reference matrix.
+ */
+BOOST_AUTO_TEST_CASE(KFNRefModelTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+
+  mlpackMain();
+
+  // Input pre-trained model.
+  SetInputParam("input_model", 
+      std::move(CLI::GetParam<KFNModel*>("output_model")));
+  
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/*
+ * Check that we can't pass an invalid tree type.
+ */
+BOOST_AUTO_TEST_CASE(KFNInvalidTreeTypeTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+  SetInputParam("tree_type", (string) "min-rp"); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/*
+ * Check that we can't pass an invalid algorithm.
+ */
+BOOST_AUTO_TEST_CASE(KFNInvalidAlgoTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+  SetInputParam("algorithm", (string) "triple_tree"); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/*
+ * Check that we can't pass an invalid value of epsilon.
+ */
+BOOST_AUTO_TEST_CASE(KFNInvalidEpsilonTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+  SetInputParam("epsilon", (double) -1); // Invalid.
```

Comment:
It's probably also worthwhile to either add another test or add another call to `mlpackMain()` where we check that `epsilon` is not 1 or greater.  (You are probably already thinking this based on the other comment I left about KFN approximation, but I thought it was a good idea to write it down so we don't forget about it. :))